Sources/MatchingEngine/Builder.swift

```diff
@@ -1,12 +1,12 @@
 import Util
 
-extension Program where Element: Hashable {
-  public typealias Predicate = (Element) -> Bool
+extension Program where Input.Element: Hashable {
+  public typealias Predicate = (Input, Input.Index) -> Input.Index?
```

Comment:
This is the consumer interface. Did you think about `(Input, in: Range<Input.Index>) -> Input.Index?`? That would allow us to avoid slicing problems, at least here.

Reply:
Good point. Change applied.